### PR TITLE
[WFLY-11480]: Possible to remove referenced activemq http-connector.

### PR DESCRIPTION
**Pull request by ehsavoie**
* Adding support for capabilities in connection factories and transport configurations.
 * Adding support for proper default configuration for external pcf.

Jira: https://issues.jboss.org/browse/WFLY-11480

### CODE REVIEW

**Comment by rsvoboda** [comment]
@mnovak1 @pkremens fyi

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/jms/ExternalPooledConnectionFactoryAdd.java [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+        final BindInfo bindInfo = ContextNames.bindInfoFor(jndiNames.get(0));
+        List<String> jndiAliases;
+        if(jndiNames.size() > 1) {
+            jndiAliases = new ArrayList<>(jndiNames.subList(1, jndiNames.size()));
+        } else {
+            jndiAliases = Collections.emptyList();
+        }
+        String managedConnectionPoolClassName = null;
+        if (resolvedModel.hasDefined(ConnectionFactoryAttributes.Pooled.MANAGED_CONNECTION_POOL.getName())) {
+            managedConnectionPoolClassName = resolvedModel.get(ConnectionFactoryAttributes.Pooled.MANAGED_CONNECTION_POOL.getName()).asString();
```
Comment (bstansberry) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/jms/ExternalPooledConnectionFactoryAdd.java]: asStringOrNull

No need to update this for my 'OrNull' comments unless you are making other changes.

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/jms/ExternalPooledConnectionFactoryAdd.java [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        }
+        final int minPoolSize = resolvedModel.get(ConnectionFactoryAttributes.Pooled.MIN_POOL_SIZE.getName()).asInt();
+        final int maxPoolSize = resolvedModel.get(ConnectionFactoryAttributes.Pooled.MAX_POOL_SIZE.getName()).asInt();
+        Boolean enlistmentTrace = null;
+        if (resolvedModel.hasDefined(ConnectionFactoryAttributes.Pooled.ENLISTMENT_TRACE.getName())) {
+            enlistmentTrace = resolvedModel.get(ConnectionFactoryAttributes.Pooled.ENLISTMENT_TRACE.getName()).asBoolean();
```
Comment (bstansberry) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/jms/ExternalPooledConnectionFactoryAdd.java]: asBooleanOrNull

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/jms/ExternalPooledConnectionFactoryAdd.java [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        }
+
+        final String txSupport;
+        if(resolvedModel.hasDefined(ConnectionFactoryAttributes.Pooled.TRANSACTION.getName())) {
+            String txType = resolvedModel.get(ConnectionFactoryAttributes.Pooled.TRANSACTION.getName()).asString();
```
Comment (bstansberry) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/jms/ExternalPooledConnectionFactoryAdd.java]: asStringOrNull

**Review by bstansberry — COMMENTED** [review]
Please add a link to the wildfly-capabilities PR describing these. (I confess I've neglected to do those for some of my recent work, but it's a new year, time to reform!)

This looks ok but I'd like a review from @jmesnil.

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/MessagingSubsystemParser_6_0.java [hunk 4]
```diff
@@ -161,6 +162,71 @@ public PersistentResourceXMLDescription getParserDescription(){
                                                         ConnectionFactoryAttributes.Pooled.INITIAL_CONNECT_ATTEMPTS,
                                                         ConnectionFactoryAttributes.Pooled.STATISTICS_ENABLED);
 
+        final PersistentResourceXMLBuilder externalPooledConnectionFactory =
```
Comment (jmesnil) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/MessagingSubsystemParser_6_0.java]: outside of the path element, is there any difference in the attributes exposed by the `external` variant of the pooled-connection-factory?

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/jms/ExternalJMSQueueDefinition.java [hunk 5]
```diff
@@ -51,7 +52,11 @@ private ExternalJMSQueueDefinition() {
 
     @Override
     public void registerAttributes(ManagementResourceRegistration registry) {
-        registry.registerReadWriteAttribute(CommonAttributes.DESTINATION_ENTRIES, null, new ReloadRequiredWriteAttributeHandler(CommonAttributes.DESTINATION_ENTRIES));
+        if (registerRuntimeOnly) {
+            registry.registerReadOnlyAttribute(CommonAttributes.DESTINATION_ENTRIES, null);
```
Comment (jmesnil) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/jms/ExternalJMSQueueDefinition.java]: change is fine (JMS resources from deployment are read-only) but how is it related to the connector/acceptor capabilities change?

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/jms/ExternalJMSTopicDefinition.java [hunk 6]
```diff
@@ -54,7 +55,11 @@ private ExternalJMSTopicDefinition() {
 
     @Override
     public void registerAttributes(ManagementResourceRegistration registry) {
-        registry.registerReadWriteAttribute(CommonAttributes.DESTINATION_ENTRIES, null, new ReloadRequiredWriteAttributeHandler(getAttributes()));
+        if (registerRuntimeOnly) {
```
Comment (jmesnil) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/jms/ExternalJMSTopicDefinition.java]: same as above

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/jms/ExternalPooledConnectionFactoryAdd.java [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+public class ExternalPooledConnectionFactoryAdd extends AbstractAddStepHandler {
+
+    public static final ExternalPooledConnectionFactoryAdd INSTANCE = new ExternalPooledConnectionFactoryAdd();
+
+    private ExternalPooledConnectionFactoryAdd() {
+        super(getDefinitions(ExternalPooledConnectionFactoryDefinition.ATTRIBUTES));
+    }
+
+    @Override
+    protected void performRuntime(OperationContext context, ModelNode operation, Resource resource) throws OperationFailedException {
```
Comment (jmesnil) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/jms/ExternalPooledConnectionFactoryAdd.java]: isn't it possible to share most of this code with PooledConnectionFactoryAdd and only specify what is actually different?

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/jms/ExternalPooledConnectionFactoryDefinition.java [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+import org.wildfly.extension.messaging.activemq.MessagingExtension;
+import org.wildfly.extension.messaging.activemq.jms.ConnectionFactoryAttributes.Common;
+import org.wildfly.extension.messaging.activemq.jms.ConnectionFactoryAttributes.Pooled;
+
+/**
+ * JMS pooled Connection Factory resource definition.
+ *
+ * @author <a href="http://jmesnil.net">Jeff Mesnil</a> (c) 2012 Red Hat Inc.
+ */
+public class ExternalPooledConnectionFactoryDefinition extends PersistentResourceDefinition {
```
Comment (jmesnil) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/jms/ExternalPooledConnectionFactoryDefinition.java]: isn't there a relation with the PooledConnectionFactoryDefinition (or a common base definition)?

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/jms/ExternalPooledConnectionFactoryRemove.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+import org.wildfly.extension.messaging.activemq.MessagingServices;
+import org.jboss.as.naming.deployment.ContextNames;
+import org.jboss.dmr.ModelNode;
+import org.jboss.msc.service.ServiceName;
+
+/**
+ *
+ * @author Emmanuel Hugonnet (c) 2018 Red Hat, inc.
+ */
+public class ExternalPooledConnectionFactoryRemove extends AbstractRemoveStepHandler {
```
Comment (jmesnil) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/jms/ExternalPooledConnectionFactoryRemove.java]: should be very similar to `PooledConnectionFactoryRemove`, could there be a shared class to handle both remove operations?

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/jms/ExternalJMSQueueDefinition.java (hunk 5 above)
Comment (ehsavoie) [reply messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/jms/ExternalJMSQueueDefinition.java]: I detected the issue (aka missing runtime description) while fixing this. Maybe i could split the commit in 2

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/MessagingSubsystemParser_6_0.java (hunk 4 above)
Comment (ehsavoie) [reply messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/MessagingSubsystemParser_6_0.java]: default values and capabilities

**Review by jmesnil — APPROVED** [review]

**Pull request merged**